torchao/prototype/mx_formats/mx_tensor.py

```diff
@@ -0,0 +1,411 @@
+    else:
+        raise AssertionError("unsupported")
+    scale_e8m0_unbiased = largest_p2_lt_max_abs - target_max_pow2
+
+    # Clamp to exponents that can be represented in e8m0
+    scale_e8m0_unbiased = torch.clamp(
+        scale_e8m0_unbiased, min=-E8M0_EXPONENT_BIAS, max=E8M0_EXPONENT_BIAS
+    )
+
+    # Create the biased e8m0 representation and cast it to 8 bits
+    scale_e8m0_biased = scale_e8m0_unbiased + E8M0_EXPONENT_BIAS
+    scale_e8m0_biased = scale_e8m0_biased.to(torch.uint8)
+
+    # Conversion to torch.uint8 sets NaN values to 0, fix this by
+    # explicitly setting known NaN values to 255
+    scale_e8m0_biased = torch.where(
+        torch.isnan(scale_e8m0_unbiased),
+        E8M0_EXPONENT_NAN_VAL,
+        scale_e8m0_biased,
+    )
+
+    # For now, calculate the scale in floating point.
+    # TODO(future) audit if there is a need to bit shift exponents instead.
+    scale_fp = torch.pow(
+        torch.full(max_abs.size(), 2.0, device=scale_e8m0_biased.device),
+        scale_e8m0_unbiased,
+    )
+
+    # Today, 2**-127 returns 0 in compile+inductor+triton because it is in the
+    # float32 denormal range. For now, manually adjust the fp scale. This is
+    # relevant if all of the incoming block values are zeroes.
+    # See https://github.com/pytorch/pytorch/issues/125557 for details.
+    # Note: it would be more correct to set the minimum to 2**-127, but this
+    # does not work in triton either as it looks like subnormal value handling
+    # has some gaps.  So, for now just set to the minimum normal value.
+    scale_fp = torch.clamp(scale_fp, min=F32_MIN_NORMAL)
+
+    # scale and saturated cast the data elements to max of target dtype
+    if elem_dtype == torch.float8_e4m3fn:
+        max_pos = F8E4M3_MAX
+    elif elem_dtype == torch.float8_e5m2:
+        max_pos = F8E5M2_MAX
+    elif elem_dtype == DTYPE_FP6_E2M3:
+        max_pos = F6_E2M3_MAX
+    elif elem_dtype == DTYPE_FP6_E3M2:
+        max_pos = F6_E3M2_MAX
+    elif elem_dtype == DTYPE_FP4:
+        max_pos = F4_E2M1_MAX
+    else:
+        raise AssertionError("unsupported")
+    data_lp = torch.clamp(
+        data_hp / scale_fp.unsqueeze(1), min=-1 * max_pos, max=max_pos
+    )
+    data_lp = data_lp.reshape(orig_shape)
+
+    # cast to target dtype
+    if elem_dtype in (torch.float8_e4m3fn, torch.float8_e5m2):
+        data_lp = data_lp.to(elem_dtype)
+    elif elem_dtype == DTYPE_FP6_E2M3:
+        data_lp = f32_to_f6_e2m3_unpacked(data_lp)
+    elif elem_dtype == DTYPE_FP6_E3M2:
+        data_lp = f32_to_f6_e3m2_unpacked(data_lp)
+    elif elem_dtype == DTYPE_FP4:
+        data_lp = f32_to_f4_unpacked(data_lp)
+        data_lp = pack_uint4(data_lp)
+    else:
+        raise AssertionError("unsupported")
+
+    return scale_e8m0_biased, data_lp
+
+
+def get_fp_scale(scale_e8m0):
+    s_offset = scale_e8m0.to(torch.int16) - E8M0_EXPONENT_BIAS
+    # TODO(later): it would be nice if there was a way to do the 2^x operation
```

Comment:
maybe this is helpful https://pytorch.org/docs/stable/generated/torch.ldexp.html

Reply:
makes sense!  I will punt this to a future person, this shouldn't be that important for e2e performance.